wallet/udb/txdb.go

```diff
@@ -2390,3 +2416,112 @@ func upgradeToVersion3(ns walletdb.ReadWriteBucket, chainParams *chaincfg.Params
 
 	return nil
 }
+
+// RemovePkScriptCreditsAndDebits removes all credits associated with the
+// provided pkScript as well as debits that spend those credits.
+func RemovePkScriptCreditsAndDebits(ns walletdb.ReadWriteBucket, pkScript []byte) error {
+	// Fetch all mined credits associated with the provided script.
+	credits := make(map[string]dcrutil.Amount, 0)
+	cc := ns.NestedReadBucket(bucketCredits).ReadCursor()
+	for k, v := cc.First(); k != nil; k, v = cc.Next() {
+		recK := extractRawCreditTxRecordKey(k)
+		recV := existsRawTxRecord(ns, recK)
+		index := extractRawCreditIndex(k)
+		scrLoc := fetchRawCreditScriptOffset(v)
+		scrLen := fetchRawCreditScriptLength(v)
+		scr, err := fetchRawTxRecordPkScript(recK, recV, index, scrLoc, scrLen)
+		if err != nil {
+			return err
+		}
+
+		if bytes.Compare(scr, pkScript) != 0 {
+			continue
+		}
+
+		amount, err := fetchRawCreditAmount(v)
+		if err != nil {
+			return err
+		}
+
+		credits[string(k)] = amount
+	}
+
+	// Fetch and delete debits that spend credits associated with the
+	// provided script.
+	debits := make([][]byte, 0)
+	dc := ns.NestedReadBucket(bucketDebits).ReadCursor()
+	for k, v := dc.First(); k != nil; k, v = dc.Next() {
+		ck := extractRawDebitCreditKey(v)
+		for key, _ := range credits {
+			if bytes.Compare(ck, []byte(key)) != 0 {
+				continue
+			}
+
+			debits = append(debits, k)
+		}
+	}
+
+	for _, k := range debits {
+		err := ns.NestedReadWriteBucket(bucketDebits).Delete(k)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Delete all credits associated with the provided script and update the
+	// mined wallet balance.
+	bal, err := fetchMinedBalance(ns)
+	if err != nil {
+		return err
+	}
+
+	for k, amt := range credits {
+		bal -= amt
+		err := ns.NestedReadWriteBucket(bucketCredits).Delete([]byte(k))
+		if err != nil {
+			return err
+		}
+	}
+
+	return putMinedBalance(ns, bal)
+}
+
+// RemovePkScriptUnminedCredits removes all unmined credits associated with
+// the provided pkScript.
+func RemovePkScriptUnminedCredits(ns walletdb.ReadWriteBucket, pkScript []byte) error {
```

Comment:
new functions added to udb should use walletdb.Read{Write}Tx instead of namespace buckets, so that the wallet package doesn't need to pass the correct buckets.